ttnn/cpp/ttnn/tensor/tensor_layout.hpp

```diff
@@ -0,0 +1,154 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "types.hpp"
+#include "enum_types.hpp"
+
+#include <cstddef>
+#include <ostream>
+
+namespace tt::tt_metal {
+
+class Size {
+public:
+    Size(size_t height, size_t width);
+    Size(const std::pair<size_t, size_t>& size);
+    Size(const std::array<size_t, 2>& size);
+
+    operator std::pair<size_t, size_t>() const;
+    operator std::array<size_t, 2>() const;
+    operator std::array<uint32_t, 2>() const;
+
+    Size operator/(const Size& rhs) const;
+    Size operator*(size_t scalar) const;
+
+    // comparison operator
+    bool operator==(const Size& rhs) const;
+
+    size_t height() const;
+    size_t width() const;
+
+    // does not have to be a member, but it is easier to find if it is
+    Size aligned_to_tile(const Size& tile);
+
+private:
+    size_t mHeight = 0;
+    size_t mWidth = 0;
+};
+
+std::ostream& operator<<(std::ostream& os, const tt::tt_metal::Size& size);
+
+// Creating as a class to differentiate the use between LegacyPadding represented as SimpleShape and Alignment.
+// This class has to eventually become its own class
+class Alignment : public ttnn::SimpleShape {
+public:
+    using ttnn::SimpleShape::SimpleShape;
+    size_t size() const { return rank(); }
+
+};
+
+std::ostream &operator<<(std::ostream &os, const tt::tt_metal::Alignment &value) {
+    os << "Alignment([";
+    for (size_t i = 0; i < value.size(); ++i) {
+        if (i > 0) {
+            os << ", ";
+        }
+        os << value[i];
+    }
+    os << "])";
+    return os;
+}
+
+
+// Alignment is a physical row alignment for each dimension of the tensor (except innermost dimension - there its a column alignment).
+// Example:
+//  Given Tensor with shape (2, 3, 10, 16) and Alignment (A:768, B:300, C:32, 32) we will physically allocate
+//     W = round_up(16, 32) = 32
+//     H_R1 = round_up(10, C:32) = 32
+//     H_R2 = round_up(R1:32 * 3, B:300) = round_up(96, 300) = 300
+//     H_R3 = = round_up(R2:300 * 2, A:768) = round_up(600, 768) = 768
+//     PHY = [H:768, W:32]
```

Comment:
I don't get why we have to store this information. The values represented in alignment is weird to me. It's almost like physical strides except the row dims don't include the width.